api/metric/api.go

```diff
@@ -22,60 +22,156 @@ import (
+
+	// GetInt64Cumulative returns a int64 cumulative TimeSeries
+	// corresponding to the provided label set.
+	GetInt64Cumulative(context.Context, *Int64CumulativeTimeSeries, ...core.KeyValue) Int64Cumulative
+
+	// GetInt64Measure returns a int64 cumulative TimeSeries
+	// corresponding to the provided label set.
+	GetInt64Measure(context.Context, *Int64MeasureTimeSeries, ...core.KeyValue) Int64Measure
+
+	// RecordBatch atomically records a batch of measurements
+	RecordBatch(context.Context, ...Measurement)
+}
+
+type Metric interface {
+	// Description returns a copy of the underlying variable.
+	Description() TimeSeries
 }
 
+// Float64Gauge is the interface to a float64 gauge TimeSeries.
 type Float64Gauge interface {
+	Metric
+
+	// Set replaces the current value of a timeseries with `value`.
+	// `labels` are additional labels applied at the call site.
 	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+
+	// Add adds to the current value of a timeseries with `value`.
+	// `labels` are additional labels applied at the call site.
+	Add(ctx context.Context, value float64, labels ...core.KeyValue)
+}
+
+// Float64Cumulative is the interface to a float64 cumulative TimeSeries.
+type Float64Cumulative interface {
+	Metric
+
+	// Inc adds to the current value of a timeseries with `value`.
+	// `value` must be >= 0.
+	// `labels` are additional labels applied at the call site.
+	Inc(ctx context.Context, value float64, labels ...core.KeyValue)
+}
+
+// Float64Measure is the interface to a float64 measure TimeSeries.
+type Float64Measure interface {
+	Metric
+
+	// Record adds a `value` observation to the timeseries.
+	// `labels` are additional labels applied at the call site.
+	Record(ctx context.Context, value float64, labels ...core.KeyValue)
 }
 
-type Handle struct {
+type Int64Gauge interface {
+	Metric
+
+	// Set replaces the current value of a timeseries with `value`.
+	// `labels` are additional labels applied at the call site.
+	Set(ctx context.Context, value int64, labels ...core.KeyValue)
+
+	// Add adds to the current value of a timeseries with `value`.
+	// `labels` are additional labels applied at the call site.
+	Add(ctx context.Context, value int64, labels ...core.KeyValue)
+}
+
+type Int64Cumulative interface {
+	Metric
+
+	// Inc adds to the current value of a timeseries with `value`.
+	// `value` must be >= 0.
+	// `labels` are additional labels applied at the call site.
+	Inc(ctx context.Context, value int64, labels ...core.KeyValue)
+}
+
+type Int64Measure interface {
+	// Record adds a `value` observation to the timeseries.
+	// `labels` are additional labels applied at the call site.
+	Record(ctx context.Context, value int64, labels ...core.KeyValue)
+}
+
+type TimeSeries struct {
 	Variable registry.Variable
 
-	Type MetricType
+	Kind Kind
 	Keys []core.Key
 }
 
-type Option func(*Handle, *[]registry.Option)
+type Measurement struct {
+	Metric Metric
+	Value  float64
```

Comment:
In other languages, I've seen the SDK convert int->float in the exporter. If we must distinguish these types and not convert int->float->int, this will have to change.